CODING.md

```diff
@@ -0,0 +1,97 @@
+# Coding guide
+
+This project is written in Go and adheres to the accepted coding styles and practices defined in [Effective Go](https://golang.org/doc/effective_go.html) and [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments). 
+
+## Static code analysis
+
+Linting is done by using https://github.com/golangci/golangci-lint with configuration defined in the root of the project.
+
+## Testing
+
+It is advisable to use a separate test package to write tests (e.g. with the `package somepkg_test` namespace). This allows explicit testing on exported behaviour of the package, not the internal implementation (also known as black-box-testing). In need to export only to testing package, `export_test.go` file should be used. Testing internal implementations is discouraged, but not forbidden. It is preferred to treat exported package behaviours as units that need to be tested.
+
+Executing `make test` should pass without any warnings or errors.
+
+## Packages
+
+Go packages with the spacemesh project should have a single and well defined responsibility, clear exposed API and tests that cover expected behaviour. To ensure better modularity of the codebase, every package should be treated as a library that provides a specific functionality and that can be used as a module in other applications. Every package should have a well written godoc page which should be used as the entry point for understanding the package's responsibility. The same as using any other third-party package. If the package godoc is not clear and requires looking at the code to understand behaviour that it provides, documentation should be improved.
+
+## Concurrency
+
+For every goroutine that is created, one must define how and when it terminates.
+
+Every channel must have an owning goroutine. That goroutine is the only one that can send to the channel, close it or transfer ownership to another goroutine. Wherever is possible a readonly or writeonly channels should be used.
+
+## Error propagation, annotation and handling
+
+Errors must be propagated by the caller function. Error should not be logged and passed at the same time. It is up to the next caller function to decide what should be done in case of an error (e.g. execute specific logic, log, or continue bubbling the error up the stack).
+
```

Comment:
I think that using `errors.Is/As` should be encouraged yes. About using switches - I'm not 100% sure because it's a general deviation from the usual `if err := ... {}` structure and switches have the problematic aspect of when they have the `default:` case it would handle the happy path which is a bit misleading. Ideally, the happy path should never be nested, but rather handled at the base nesting level of the method (only error paths should be nested). It is indeed sometimes unavoidable to handle happy path this way but switches may cause this to happen more often than not.

Also as a sidenode it should probably also be noted that happy path shouldn't be nested :)

I'd like to cover all those in a separate styleguide that would have examples too.

Reply:
Using switch doesn't mean that the happy path must be nested. The errors can be handled in the switch cases and the happy path - below: 
```go
err := someFunctionCall()
switch {
case errors.Is(err, Error1):
   // handle Error1
case err != nil:
   // handle unknown error
}

// happy path goes here
```